### PR TITLE
TUI 'autoread'

### PR DESCRIPTION
**Pull request by BK1603**
Continuation of #5599
I couldn't save the authorship of @justinmk as the PR was too old and resulted in a lot of merge conflicts.
The issues faced by justin seem to have been fixed through other unrelated PRs over time.

Tests remaining. 
@teto

### CODE REVIEW

File: src/nvim/aucmd.c [hunk 1]
```diff
@@ -50,12 +52,46 @@ void aucmd_schedule_focusgained(bool gained)
...
-  apply_autocmds((gained ? EVENT_FOCUSGAINED : EVENT_FOCUSLOST),
-                 NULL, NULL, false, curbuf);
+  need_redraw |= apply_autocmds((gained ? EVENT_FOCUSGAINED : EVENT_FOCUSLOST),
+                                NULL, NULL, false, curbuf);
+
+  // When activated: Check if any file was modified outside of Vim.
+  // Only do this when not done within the last two seconds (could get
+  // several events in a row).
+
+  if (gained && last_time + (Timestamp)2 < os_time()) {
```
Comment (teto) [inline src/nvim/aucmd.c:69]: let's make it more granular, instead use `os_now` that returns milliseconds and check against a delta of 200ms for instance.

Comment (teto) [inline src/nvim/aucmd.c:69]: we can keep the time check. because the filesystem timestamps seem to be in seconds, it makes no sense to check between less than a second but I think we could use os_now for moree precision and maybe check around 1500ms.

Comment (BK1603) [reply src/nvim/aucmd.c:69]: Sure thing!

File: test/functional/autoread/focus_spec.lua [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      {1: }                                                 |
+      {4:~                                                 }|
+      {4:~                                                 }|
+      {4:~                                                 }|
+      {5:xtest-foo                                         }|
+      :edit xtest-foo                                   |
+      {3:-- TERMINAL --}                                    |
+    ]]}
+
+    local file = io.open(path, 'w')
```
Comment (teto) [inline test/functional/autoread/focus_spec.lua:47]: you can use helpers.writefile to fill the file with content. It takes a flag to append too

File: src/nvim/aucmd.c [hunk 3]
```diff
@@ -50,12 +52,46 @@ void aucmd_schedule_focusgained(bool gained)
...
+  need_redraw |= apply_autocmds((gained ? EVENT_FOCUSGAINED : EVENT_FOCUSLOST),
+                                NULL, NULL, false, curbuf);
+
+  // When activated: Check if any file was modified outside of Vim.
+  // Only do this when not done within the last two seconds (could get
+  // several events in a row).
+
+  if (gained && last_time + (Timestamp)1500 < os_now()) {
+    need_redraw = check_timestamps(true);
+    last_time = os_time();
```
Comment (teto) [inline src/nvim/aucmd.c:71]: this should be os_now() too

Comment (BK1603) [reply src/nvim/aucmd.c]: Done!

File: test/functional/autoread/focus_spec.lua (hunk 2 above)
Comment (BK1603) [reply test/functional/autoread/focus_spec.lua]: Done!

File: test/functional/autoread/focus_spec.lua [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      {1: }                                                 |
+      {4:~                                                 }|
+      {4:~                                                 }|
+      {4:~                                                 }|
+      {5:xtest-foo                                         }|
+      :edit xtest-foo                                   |
+      {3:-- TERMINAL --}                                    |
+    ]]}
+
+    helpers.sleep(2000)
```
Comment (teto) [inline test/functional/autoread/focus_spec.lua:47]: I think here you can sleep less than 2000 milliseconds, say 1000.

File: src/nvim/aucmd.c [hunk 5]
```diff
@@ -50,12 +52,46 @@ void aucmd_schedule_focusgained(bool gained)
...
   recursive = true;
-  apply_autocmds((gained ? EVENT_FOCUSGAINED : EVENT_FOCUSLOST),
-                 NULL, NULL, false, curbuf);
+  need_redraw |= apply_autocmds((gained ? EVENT_FOCUSGAINED : EVENT_FOCUSLOST),
+                                NULL, NULL, false, curbuf);
+
+  // When activated: Check if any file was modified outside of Vim.
+  // Only do this when not done within the last two seconds (could get
+  // several events in a row).
+
```
Comment (teto) [inline src/nvim/aucmd.c:68]: remove the empty line, also mention that it makes sense to wait at least one second since on some filesystems the timestamp granularity is in second.

Comment (jamessan) [reply src/nvim/aucmd.c:68]: FAT32 has a 2 second granularity for modification time.  Most contemporary file systems are much more granular.

File: test/functional/autoread/focus_spec.lua [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+local feed_data = thelpers.feed_data
+
+if helpers.pending_win32(pending) then return end
+
+describe('autoread TUI FocusGained/FocusLost', function()
+  local screen
+
+  before_each(function()
+    clear()
+    screen = thelpers.screen_setup(0, '["'..nvim_prog
```
Comment (jamessan) [inline test/functional/autoread/focus_spec.lua:16]: Can't you just use `screen = thelpers.screen_setup()` here?

File: test/functional/autoread/focus_spec.lua [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+      ..'", "-u", "NONE", "-i", "NONE", "--cmd", "set noswapfile noshowcmd noruler"]')
+    feed_command('set autoread')
+    feed_data("\034\016")
```
Comment (jamessan) [inline test/functional/autoread/focus_spec.lua:19]: Why is this needed?

File: test/functional/autoread/focus_spec.lua [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    local expected_addition = [[
+    line 1
+    line 2
+    line 3
+    line 4
+    ]]
+
+    helpers.write_file(path, '')
+    feed_command('edit '..path)
+    retry(2, 3 * screen.timeout, function()
```
Comment (jamessan) [inline test/functional/autoread/focus_spec.lua:33]: Why do we need retry here?

File: test/functional/autoread/focus_spec.lua (hunk 4 above)
Comment (jamessan) [reply test/functional/autoread/focus_spec.lua:47]: I wonder if back-dating the modification time (e.g. now - 10 seconds) when the initial file is created would be better?  `lfs.touch` would allow us to do that.

This should ensure that it's updated when the new contents are added.

File: test/functional/autoread/focus_spec.lua [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+      {4:~                                                 }|
+      {5:xtest-foo                                         }|
+      :edit xtest-foo                                   |
+      {3:-- TERMINAL --}                                    |
+    ]]}
+
+    helpers.sleep(2000)
+    helpers.write_file(path, expected_addition)
+
+    retry(2, 3 * screen.timeout, function()
```
Comment (jamessan) [inline test/functional/autoread/focus_spec.lua:50]: Again, why is retry needed?

File: test/functional/autoread/focus_spec.lua (hunk 4 above)
Comment (BK1603) [reply test/functional/autoread/focus_spec.lua:47]: This sounds good. We wouldn't have to wait during the test that way.

File: test/functional/autoread/focus_spec.lua (hunk 7 above)
Comment (BK1603) [reply test/functional/autoread/focus_spec.lua:19]: You're right. I don't think we need the `feed_command` or `feed_data` lines here. I will remove them in the next commit :)

File: test/functional/autoread/focus_spec.lua (hunk 8 above)
Comment (BK1603) [reply test/functional/autoread/focus_spec.lua:33]: This is something that I saw in the [tui_spec.lua#L914](https://github.com/neovim/neovim/blob/master/test/functional/terminal/tui_spec.lua#L914) file and thought that it must be because the function `feed_data` might not always be successful. I remember reading that 0 can be returned  (via chansend) for the cases when no bytes are written. So I thought we retry to ensure that the specified key was received by the child nvim instance. 

Although in that case I think I should leave the expect calls inside the retry block. I shouldn't have moved them out.

File: test/functional/autoread/focus_spec.lua (hunk 6 above)
Comment (BK1603) [reply test/functional/autoread/focus_spec.lua:16]: I don't think I can, since we open a child neovim instance and trigger the `FocusGained` and `FocusLost` event in this child instance, and see if the file was reloaded. The defualt command for `screen_setup` is `nvim_dir/tty-test`. I wasn't too sure of whether this is what was required and also saw that this was what was already being used in [tui_spec.lua#L906](https://github.com/neovim/neovim/blob/master/test/functional/terminal/tui_spec.lua#L906) so decided using this.

File: test/functional/autoread/focus_spec.lua (hunk 8 above)
Comment (BK1603) [reply test/functional/autoread/focus_spec.lua]: I removed the retries. It works without them :)

File: xtest-foo [hunk 10]
```diff
@@ -0,0 +1,4 @@
+line 1
+line 2
```
Comment (teto) [inline xtest-foo:2]: we dont need this file

Comment (BK1603) [reply xtest-foo]: This was a mistake, sorry :sweat_smile:

File: src/nvim/aucmd.c [hunk 11]
```diff
@@ -50,12 +52,47 @@ void aucmd_schedule_focusgained(bool gained)
...
+    need_wait_return = false;
+
+    if (State & CMDLINE) {
+      redrawcmdline();
+    } else if ((State & NORMAL) || (State & INSERT)) {
+      if (must_redraw != 0) {
+        update_screen(0);
+      }
+
+      setcursor();
```
Comment (justinmk) [inline src/nvim/aucmd.c:87]: yikes this is a mess. https://github.com/neovim/neovim/pull/5599 was WIP for a reason. https://github.com/neovim/neovim/issues/16809
